### ISSUE TITLE
Support `ruff check --fix` as a formatter

### ISSUE THREAD

**Issue by zmeir** [akaihola/darker#758]
#744 will add support for `ruff format` as a code-formatter.

Ruff also supports the [`ruff check --fix`](https://docs.astral.sh/ruff/linter/#fixes) in its linter command, that allows it to automatically fix some linter errors. This essentially turns `ruff check` from a linter to a code-formatter.

It would be nice if we could also use `ruff check --fix` as a code-formatter with Darker.  
There are also additional flags to control the auto-fixing behavior, that maybe need to be taken into account: `--fix-only`, `--unsafe-fixes`.

> Why not just use [Graylint](https://github.com/akaihola/Graylint)?

Graylint is a perfect solution for running linter commands that __do not modify the code__.  
However, with `ruff check --fix`, the "linter" will automatically modify the code the fix the linter errors. This means that `ruff check --fix` is not only a linter but also a code-formatter, and as such is incompatible with Graylint.

**Comment by akaihola** [comment]
@zmeir, there's one complication we need to think about.

The algorithm Darker uses to match code changes from 1) the user and 2) reformatting is not foolproof. It does a best effort of comparing and matching diff chunks, but in some edge cases it will get confused and reassemble the Python source code file in the wrong way.

To mitigate this, Darker always compares the AST before and after applying the reformatted chunks, and if the AST didn't stay intact, it will run the whole diff matching again but enlarging the chunks considered to have been edited by the user. This process is repeated until the AST matches, or until the entire file is considered as edited and thus all reformatting changes are applied.

Now, without investigation and testing, it's impossible to know whether the changes done by `ruff check --fix` will trigger such an edge case. One thing is sure though: the AST won't be unchanged after ruff's fixes, so AST verification can't be used to mitigate mistakes done by Darker's diff matching algorithm.

So investigation, testing and considering different mitigation strategies is needed.

For earlier discussions on the diff algorithm, see:
- [Darker changes lines when the previous lines got changed #388](/akaihola/darker/issues/388#issuecomment-1230707491)
- #206
- #204
- #371 
- [DRAFT: Apply pyupgrade. #51](/akaihola/darker/pull/51#discussion_r471931014)
- #20

**Comment by zmeir** [comment]
Thanks for the response @akaihola !

I was aware of the AST verification Darker does, but it hadn't occurred to me that `ruff check --fix` will modify the AST until you pointed it out. Indeed, rewriting code is more than just formatting...

I'll try to play around with `graylint --lint="ruff check"` and see if I can use the output to construct a command to `ruff check --fix-only` that will only target the specific issues that were reported in the diff.

**Comment by akaihola** [comment]
One obvious (but currently impossible) way to solve this would be to somehow have `ruff check --fix` report only the `N`th fix, with `N` being accepted on the command line.

Darker could then start with fix 1, and only apply it if the lines it changes intersect with any of the user modified lines.

Darker would track changed line numbers of user modifications (we have code for this in Graylint), and proceed to fix 2 (if fix 1 was _not_ applied) or the new fix 1 (if the original fix 1 _was_ applied).

Unfortunately it's unrealistic to expect `ruff` to implement such a feature. The performance penalty might also be noticeable.

**Comment by akaihola** [comment]
I'm experimenting with a [pyupgrade] formatter plugin in #755, and it has exactly the same problem. Currently I'm just sidestepping the issue by disabling AST verificating and enforcing applying all pyupgrade changes to all user-modified files.

We could of course make it one notch smarter by only applying all changes to the file if at least one changed line intersects with user modifications.

[pyupgrade]: /asottile/pyupgrade

**Comment by akaihola** [comment]
I've also pondered whether running a non-AST-preserving formatter first for the old, unmodified version of the code (e.g. `main`) and then again for the modified one (e.g. feature branch or working tree) would give enough information to figure out how to isolate changes touching the modified lines.

But it becomes rather complicated with many moving parts, and I suspect this approach won't fly either.

**Comment by akaihola** [comment]
> I'll try to play around with `graylint --lint="ruff check"` and see if I can use the output to construct a command to `ruff check --fix-only` that will only target the specific issues that were reported in the diff.

Do you mean adding `--fix-only` and `--select` with a comma-separated list of specific rules?

That could indeed be another way to increase granularity – probably in most cases it would help only apply changes touching modified lines.

But the case in which it fails is when the same fix is applied both to lines which have been modified by the user and to lines which haven't. In that case I'm afraid the best we can do is to apply all instances of the same fix inside a file.

**Comment by zmeir** [comment]
> Do you mean adding --fix-only and --select with a comma-separated list of specific rules?

Yes. I was hoping to find other ruff configurations to help me narrow the scope even more, like an option to limit the line range to check, but I found nothing like this.

> One obvious (but currently impossible) way to solve this would be to somehow have ruff check --fix report only the Nth fix, with N being accepted on the command line.

Obvious but impossible. That sums up what I'm looking for :)

> I'm experimenting with a [pyupgrade](https://github.com/asottile/pyupgrade) formatter plugin in https://github.com/akaihola/darker/pull/755 [...] We could of course make it one notch smarter by only applying all changes to the file if at least one changed line intersects with user modifications.

That sounds like a great solution. This combined with some narrowing of rules using `--select` sounds like it could work very well.

> I've also pondered whether running a non-AST-preserving formatter first for the old, unmodified version of the code (e.g. main) and then again for the modified one (e.g. feature branch or working tree) would give enough information to figure out how to isolate changes touching the modified lines.

I had a similar thought. What if we do something like this?
1. Run `ruff check --fix` on the unmodified code
2. Run `ruff check --fix` on the modified code
3. Save the diff before running step (2) on the modified code - call it "branch diff"
4. Save the diff between the result of (1) and (2) - call it "fix diff"
5. Keep any diff chunks from "branch diff" that intersect with at least one chunk in "fix diff"

I'm sure there are holes in this flow. Just a thought...

**Comment by akaihola** [comment]
Oops @zmeir, I didn't mean to edit your comment. I must have selected the wrong drop-down menu item when I was going to do `Quote reply`. I wonder if I can somehow recover the original content...
